game/src/sandbox/dashboards/trip_table.rs

```diff
@@ -4,52 +4,177 @@ use abstutil::prettyprint_usize;
+                }
+            ))
+            .tooltip(Text::from(Line("Finished Trips")));
+
+        let finished_trips_table = make_table_finished_trips(app);
+        let finished_trips_content = Widget::col(vec![
+            finished_trips_table.render(ctx, app),
+            Filler::square_width(ctx, 0.15)
+                .named("preview")
+                .centered_horiz(),
+        ]);
+        tabs.push_tab(finished_trips_btn, finished_trips_content);
+
+        let cancelled_trips_table = make_table_cancelled_trips(app);
+        let cancelled_trips_btn = ctx
+            .style()
+            .btn_tab
+            .text(format!("{} Cancelled Trips", prettyprint_usize(cancelled)))
+            .tooltip(Text::from(Line("Cancelled Trips")));
+        let cancelled_trips_content = Widget::col(vec![
+            cancelled_trips_table.render(ctx, app),
+            Filler::square_width(ctx, 0.15)
+                .named("preview")
+                .centered_horiz(),
+        ]);
+        tabs.push_tab(cancelled_trips_btn, cancelled_trips_content);
+
+        let unfinished_trips_table = make_table_unfinished_trips(app);
+        let unfinished_trips_btn = ctx
+            .style()
+            .btn_tab
+            .text(format!(
+                "{} ({:.1}%) Unfinished Trips",
+                prettyprint_usize(unfinished),
+                if total > 0 {
+                    (unfinished as f64) / (total as f64) * 100.0
+                } else {
+                    0.0
+                }
+            ))
+            .tooltip(Text::from(Line("Unfinished Trips")));
+        let unfinished_trips_content = Widget::col(vec![
+            unfinished_trips_table.render(ctx, app),
+            Filler::square_width(ctx, 0.15)
+                .named("preview")
+                .centered_horiz(),
+        ]);
+        tabs.push_tab(unfinished_trips_btn, unfinished_trips_content);
+
+        let panel = Panel::new(Widget::col(vec![
+            DashTab::TripTable.picker(ctx, app),
+            tabs.build_widget(ctx),
+        ]))
+        .exact_size_percent(90, 90)
+        .build(ctx);
+
+        Self {
+            tab: DashTab::TripTable,
+            table_tabs: tabs,
+            panel,
+            finished_trips_table,
+            cancelled_trips_table,
+            unfinished_trips_table,
+        }
     }
 }
 
-pub struct UnfinishedTripTable;
-
-impl UnfinishedTripTable {
-    pub fn new(ctx: &mut EventCtx, app: &App) -> Box<dyn State<App>> {
-        GenericTripTable::new(
-            ctx,
-            app,
-            DashTab::UnfinishedTripTable,
-            make_table_unfinished_trips(app),
-            make_panel_unfinished_trips,
-        )
+impl State<App> for TripTable {
+    fn event(&mut self, ctx: &mut EventCtx, app: &mut App) -> Transition {
+        match self.panel.event(ctx) {
+            Outcome::Clicked(x) => {
+                if self.table_tabs.active_tab_idx() == 0 && self.finished_trips_table.clicked(&x) {
```

Comment:
I really like the improvements to the code here! Using `push_tab` to build them up and keeping all of the tabs "alive" at all times is quite intuitive. There's a little bit of boilerplate to associate the tab index with the content and route events, but I think it's very localized and easy to understand. The alternatives are likely much more bizarre to express in Rust, with callbacks or implementing some complicated trait. So I like this!

Looking forward to applying this to info panels in particular, there might be some more complications with conceptually keeping all tabs' content "active" at the same time. Some tabs auto-update as time passes, and this can be a bit expensive, so we wouldn't want to necessarily do it for multiple tabs. Also they fill out this weird `Details` struct to draw extra zoomed/unzoomed stuff on the map, and it wouldn't make sense to draw multiple tabs' extra content at once. Maybe we just wind up using the new tab button style there, but for the moment, leave the existing weird structures in place for updating/switching tabs.